File: src/content/docs/topics/cs/coding-problems/stack/394-decode-string-approach2.py
def decode_string(s: str) -> str:
    stack = []
    current_string = ""
    current_count = 0

    for ch in s:
        if ch.isdigit():
            current_count = current_count * 10 + int(ch)
        elif ch == "[":
            stack.append((current_count, current_string))
            current_count = 0
            current_string = ""
        elif ch == "]":
            count, prefix = stack.pop()
            current_string = prefix + count * current_string
        else:
            current_string += ch

    return current_string

assert decode_string("3[a]2[bc]") == "aaabcbc"
assert decode_string("3[a2[c]]") == "accaccacc"
assert decode_string("2[abc]3[cd]ef") == "abcabccdcdcdef"
assert decode_string("a") == "a"
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/stack/394-decode-string-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function decodeString(s: string): string {
    const stack: [number, string][] = [];
    let currentString = '';
    let currentCount = 0;

    for (const ch of s) {
        if (ch >= '0' && ch <= '9') {
            currentCount = currentCount * 10 + Number(ch);
        } else if (ch === '[') {
            stack.push([currentCount, currentString]);
            currentCount = 0;
            currentString = '';
        } else if (ch === ']') {
            const [count, prefix] = stack.pop()!;
            currentString = prefix + currentString.repeat(count);
        } else {
            currentString += ch;
        }
    }

    return currentString;
}

assert(decodeString('3[a]2[bc]') === 'aaabcbc');
assert(decodeString('3[a2[c]]') === 'accaccacc');
assert(decodeString('2[abc]3[cd]ef') === 'abcabccdcdcdef');
assert(decodeString('a') === 'a');
console.log('all tests pass');
